[WIP] qt57: add darwin support + add new modules

adolfogc opened the pull request:
###### Main changes
- Add support for Mac OS X (currently only Qt4 version 4.8 is available for Darwin).
- Add new Qt modules: QtMacExtras, QtQuickControls2, QtSCXML, QtSerialBus.
- Other minor changes:
  - Modify _maintainers/scripts/generate-qt.sh_ to fetch tarballs using TLS.
  - Change _substituteInPlace_ instances for _grep_ (otherwise it fails building in Darwin).
  - Build QtMultimedia with _openal_ support.
  - Build QtImageFormats with _libmng_, _libwebp_ and _jasper_ support.
  - Build Qt using systems _double-conversion_ library.
  - Build QtConnectivity with bluez support (Linux).
  - Remove line that fixed bison to bison2. It was a workaround to a bug that was fixed with Qt 5.2.0 ([see this](https://bugreports.qt.io/browse/QTBUG-32913)).
###### TO-DO
- [ ] Fix issue causing regression in Nixos.
- [ ] Figure out why some _dylib_s aren't found by _ld_ (Darwin).
- [ ] Figure out why the Travis build failed for Xcode:
  
  `The invocation of "nix-build -A qt57.qtdoc -A qt57.libdbusmenu -A qt57.qca-qt5 -A qt57.full -A qt57.qtconnectivity -A qt57.qtquickcontrols2 -A qt57.qtmacextras -A qt57.poppler -A qt57.qtmultimedia -A qt57.qtscript -A qt57.qtgraphicaleffects -A qt57.libcommuni -A qt57.qtxmlpatterns -A qt57.qtsensors -A qt57.accounts-qt -A qt57.qtscxml -A qt57.qtsvg -A qt57.qtquickcontrols -A qt57.grantlee`
- [ ] ~~Add _qtstyleplugins_ as a library separate from Qt5.~~
- [ ] Check for [#17413](https://github.com/NixOS/nixpkgs/pull/17413). A rebase and a fix may be needed if it is merged before this one.
- [ ] Check for [#16561](https://github.com/NixOS/nixpkgs/pull/16561). It enables building the QtWebEngine module in Qt 5.6.
###### Things done
- [X] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [X] NixOS
  - [X] OS X
  - [X] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

ttuegel commented on pkgs/development/libraries/qt-5/5.7/qtstyleplugins.nix:
```diff
@@ -0,0 +1,7 @@
+{ qtSubmodule, qtbase, pkgconfig, GConf, gtk, libX11}:
+
+qtSubmodule {
```
@adolfogc Please add this module as a library separate from Qt until it is part of a release.

ttuegel commented:
The update to Qt 5.7 is superseded by #17369, so you'll need to rebase your changes.

Thanks for getting this working on Darwin!

adolfogc commented:
@ttuegel I've rebased and included the darwin changes (please note there are other changes as well). I omitted adding the QtStylePlugins library because I saw you added support for building the GTK theme in NixOS/nixpkgs#17369 and I don't see a reason to add those plugins anymore.

ttuegel commented on pkgs/development/libraries/qt-5/5.7/qtbase/default.nix:
```diff
@@ -91,106 +77,132 @@ stdenv.mkDerivation {
-    -qml-debug
+    -rpath
+    -verbose
+    -silent
```
Using both `-verbose` and `-silent` is probably a mistake.

adolfogc replied:
Thanks for the catch, I just rebased, fixed and pushed back.

ttuegel commented on pkgs/development/libraries/qt-5/5.7/qtbase/default.nix:
```diff
@@ -91,106 +77,132 @@ stdenv.mkDerivation {
     -no-linuxfb
-    -no-kms
-
-    ${lib.optionalString (!system-x86_64) "-no-sse2"}
```
IIUC, in the absence of these flags, Qt will try to detect these features based on the host CPU. That makes builds non-deterministic and it can actually result in binary incompatibility!

adolfogc replied:
Ok thanks, didn't know that. Will fix!

ttuegel commented on pkgs/development/libraries/qt-5/5.7/qtbase/default.nix:
```diff
@@ -91,106 +77,132 @@ stdenv.mkDerivation {
+    ]
     ++ lib.optional (postgresql != null) postgresql;
 
-  nativeBuildInputs =
```
The inputs listed here should stay in `nativeBuildInputs`.

adolfogc replied:
I'd a doubt with this one, does this package gets cross-compiled?

adolfogc commented:
@ttuegel  I've made the requested changes. What's the difference between buildInputs and nativeBuildInputs when you're not cross-compiling?

ttuegel commented:
> What's the difference between buildInputs and nativeBuildInputs when you're not cross-compiling?

None, except documenting which inputs may be retained as runtime dependencies. I don't know if it will cross-compile, but I inherited the package that way.

adolfogc commented:
Ok, thanks. I was confused with how should I use those two lists for Qt.

ttuegel commented:
Is this ready to go on your end? (The title still says "[WIP]" so I wanted to double-check.)

adolfogc commented:
I still want to check why the travis build isn't working for Xcode.

adolfogc commented:
The build job run out of time. I don't yet know why, but some texlive-* derivations got triggered.
I also checked on my mac and linux test machines; some texlive-* derivations got installed.

adolfogc commented on pkgs/development/libraries/qt-5/5.7/qtbase/default.nix:
```diff
@@ -36,97 +42,65 @@ stdenv.mkDerivation {
-
-    configureFlags+="\
-        -plugindir $out/lib/qt5/plugins \
-        -importdir $out/lib/qt5/imports \
```
@ttuegel I moved these lines to configureFlags.

adolfogc commented on pkgs/development/libraries/qt-5/5.7/qtbase/default.nix:
```diff
@@ -36,97 +42,65 @@ stdenv.mkDerivation {
-        -plugindir $out/lib/qt5/plugins \
-        -importdir $out/lib/qt5/imports \
-        -qmldir $out/lib/qt5/qml \
-        -docdir $out/share/doc/qt5"
```
@ttuegel I moved these lines (86-89) to configureFlags.

adolfogc commented:
@ttuegel Please see [this discussion](NixOS/nixpkgs#4855). Should we rename nativeBuildInputs to buildInputs?

> **buildInputs**
A list of dependencies used by stdenv to set up the environment for the build. For each dependency dir, the directory dir/bin, if it exists, is added to the PATH environment variable. Other environment variables are also set up via a pluggable mechanism. For instance, if buildInputs contains Perl, then the lib/site_perl subdirectory of each input is added to the PERL5LIB environment variable. See Section 3.6, “Package setup hooks” for details.
**propagatedBuildInputs**
Like buildInputs, but these dependencies are propagated: that is, the dependencies listed here are added to the buildInputs of any package that uses this package as a dependency. So if package Y has propagatedBuildInputs = [X], and package Z has buildInputs = [Y], then package X will appear in Z’s build environment automatically.

Taken from: [https://nixos.org/nixpkgs/manual/#ssec-stdenv-attributes](https://nixos.org/nixpkgs/manual/#ssec-stdenv-attributes)

What I understand from reading the above is that _propagatedBuildInputs_ is used to speciefy the packages retained as runtime dependencies.

ttuegel commented:
> What I understand from reading the above is that propagatedBuildInputs is used to specify the packages retained as runtime dependencies.

That is not true. Runtime dependencies are determined by Nix automatically at the end of a build by searching the output for store paths. Runtime dependencies can come from any of the build inputs, whether propagated or not. When cross-compiling, they cannot come from anything `native`, because those dependencies come from the build system. 

The difference between `buildInputs` and `propagatedBuildInputs` is the following

- everything in `buildInputs` it is included in the build environment for this package
- everything in `propagatedBuildInputs` is included in the build environment for the package _and_ in the build environment for every package that depends on this one.

adolfogc commented:
Ok, thanks for explaining the difference with buildInputs and propagatedBuildInputs and their relation to runtime dependencies. Regarding the other (difference between buildInputs and nativeBuildInputs) what is the action to take. Is it a good idea to rename nativeBuildInputs to buildInputs? My understanding is that they get both combined in a single list.

bennofs commented:
@adolfogc That issue is about a greater refactoring of nixpkgs itself. For packages, the rule is as follows:

* everything that should be build for the platform where the output is run should be in `buildInputs` (so libraries the output depends on, etc)
* everything that should be build for the platform on which the *compile* happens should be in `nativeBuildInputs` (so build tools, etc)

This is what should be used for the time being. If that nixpkgs refactoring happens, then it will change, but since that is not yet the case, you should stick to the current meaning of these attributes.

ttuegel commented on pkgs/development/libraries/qt-5/5.7/qtbase/default.nix:
```diff
@@ -135,61 +109,118 @@ stdenv.mkDerivation {
 
   nativeBuildInputs =
-    [ bison flex gperf lndir patchelf perl pkgconfig python ruby ];
+  [ bison flex gperf lndir patchelf perl pkgconfig python ruby bash ];
```
@adolfogc `bash` is already included in `stdenv` so this should not be necessary.

ttuegel commented:
The build of `qt57.qtbase` fails for me on NixOS. The builder spawns more than 300 compiler processes and then dies when it runs out of memory. I can build it from master, so there seems to be a regression hidden here.

adolfogc commented:
@bennofs Thanks, your explanation and ttuegels have helped me understand how I should use those lists correctly.

adolfogc commented:
@ttuegel  Thanks for the feedback, I will look into what's causing this regression and remove the unnecesary _bash_.

adolfogc committed:
qt57: Add Darwin support and other minor changes.

- Adds support for Mac OS X (currently only Qt4 version 4.8 is available for Darwin).
- Adds new Qt modules: QtMacExtras, QtQuickControls2, QtSCXML, QtSerialBus.
- Other minor changes:
   - Modifies `maintainers/scripts/generate-qt.sh` to fetch tarballs using TLS.
   - Changes some substituteInPlace instances for grep (otherwise it fails building in Darwin).
   - Builds QtMultimedia with openal support.
   - Builds QtImageFormats with libmng, libwebp and jasper support.
   - Builds Qt using systems double-conversion library.
   - Builds QtConnectivity with bluez support (Linux).
   - Remove line that fixed bison to bison2. It was a workaround to a bug that was fixed with Qt 5.2.0.

Thanks to @ttuegel for all the feedback.

ttuegel commented:
I am closing this PR because it is outdated. Darwin support for all Qt 5 versions was added by another PR. Please open a new PR if there are still missing submodules.

adolfogc commented:
@ttuegel Thanks, good to hear Darwin support was added for Qt 5!
